### PR TITLE
Add threshold for ops using openmp macro

### PR DESCRIPTION
**Pull request by zy97140**
As described in issue #4188 and [benchmark](https://github.com/zy97140/omp-benchmark-for-pytorch), optimal OpenMP threshold is dependent on operation type, CPU type and tensors' contiguity.
This PR add a parameter named OMP_THRESHOLD for macros TH_TENSOR_APPLYX_OMP to control the OpenMP threshold.

### CODE REVIEW

File: aten/src/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -3712,6 +3726,36 @@ TENSOR_IMPLEMENT_LOGICAL_SUM(logicalany, ||, 0)
 #define TH_MATH_NAME(fn) fn
 #endif
 
+#ifdef _OPENMP
+LAB_IMPLEMENT_BASIC_FUNCTION(log,TH_MATH_NAME(log),HYPER_TH_OMP_OVERHEAD_THRESHOLD)
```
Comment (ezyang) [inline aten/src/TH/generic/THTensorMath.c]: Is there something we can do to avoid duplicating definitions for all of these pointwise operations?

**Comment by ezyang** [comment]
CC @cpuhrsch

**Comment by cpuhrsch** [comment]
Thanks @ezyang. I'm not sure what the philosophy here is, but CC @colesbury should know more.

File: aten/src/TH/generic/THTensorMath.c (hunk 1 above)
Comment (MlWoo) [reply aten/src/TH/generic/THTensorMath.c]: Maybe we have to modify the `LAB_IMPLEMENT_BASIC_FUNCTION` Macro when openmp is unavailable. How about adding a default OMP_OVERHEAD_THRESHOLD to the macro? It is a little annoying to pass a additional parameter when using the macro.

**Comment by zy97140** [comment]
@ezyang @colesbury Could you spare some time to review this PR? I want to optimize nn module  based on these macros.

File: aten/src/TH/generic/THTensorMath.c (hunk 1 above)
Comment (ezyang) [reply aten/src/TH/generic/THTensorMath.c]: An idiom which we have used in `interned_strings.h` looks something like this:

```
  #define FORALL_JIT_SYMBOLS(_) \
  _(Assign) \
  _(Constant) \
  _(CppOp) \
  _(Drop) \
  _(Eval) \
  _(Expand) \
```

and then

```
    enum BuiltinSymbol {
      #define DEFINE_SYMBOL(s) \
        k##s,
      FORALL_JIT_SYMBOLS(DEFINE_SYMBOL)
      #undef DEFINE_SYMBOL
      kLastSymbol, //where we start counting for new symbols
    };
```

It seems like you might to apply this principle here, where `LAB_IMPLEMENT_BASIC_FUNCTION` is the parameterized function, and it discards the threshold argument depending on the conditional.

**Review by ezyang — APPROVED** [review]
I'm OK with merging this but I don't regularly interact with this code.

File: aten/src/TH/generic/THTensorMath.c (hunk 1 above)
Comment (MlWoo) [reply aten/src/TH/generic/THTensorMath.c]: @ezyang [C macros with optional parameters](https://stackoverflow.com/questions/3046889/optional-parameters-with-c-macros) is a nice idea to handle with the situation more flexibly.

**Comment by zy97140** [comment]
@soumith @fmassa @ezyang @zdevito  @vedanuj @cpuhrsch Could you spare some time to review this PR? I want to optimize nn module based on these macros. Thank you very much.

File: aten/src/TH/THTensorApply.h [hunk 2]
```diff
@@ -461,7 +460,7 @@
       } \
     } else {\
       PRAGMA(simd) \
-      PRAGMA( omp parallel for if (SIZE > TH_OMP_OVERHEAD_THRESHOLD_OMP) )  \
+      PRAGMA( omp parallel for if (SIZE > OMP_THRESHOLD * 10) )  \
```
Comment (apaszke) [inline aten/src/TH/THTensorApply.h]: The magical `* 10` are weird. Why can't we just increase the threshold?

File: aten/src/TH/generic/THTensorMath.c [hunk 3]
```diff
@@ -10,6 +10,9 @@
 #include <omp.h>
 #endif
 
+#define HYPER_TH_OMP_OVERHEAD_THRESHOLD 2000
+#define ORDIN_TH_OMP_OVERHEAD_THRESHOLD 20000
+#define UNCERTAIN_TH_OMP_OVERHEAD_THRESHOLD 50000
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: Why is it "uncertain"?

File: aten/src/TH/THTensorApply.h (hunk 2 above)
Comment (MlWoo) [reply aten/src/TH/THTensorApply.h]: @apaszke we can increase the threshold directly. But we just want to follow the exsiting code. eg. light weight operations like [left shift ](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorMath.c#L924).

File: aten/src/TH/generic/THTensorMath.c (hunk 3 above)
Comment (zy97140) [reply aten/src/TH/generic/THTensorMath.c]: You can look into this [benchmark](https://github.com/zy97140/omp-benchmark-for-pytorch), in which we just tested operations including copy, add, div, sin, exp, sum, and prod on Intel(R) Xeon(R) CPU E5-2699 v4, Intel(R) Xeon(R) Platinum 8180 CPU and Intel(R) Core(TM) i7-5960X CPU. We have not covered all operations and all CPUs yet. This uncertain threshold is generally a good one according to our experience attained from the benchmarking process.

File: aten/src/TH/THTensorApply.h (hunk 2 above)
Comment (zy97140) [reply aten/src/TH/THTensorApply.h]: The input parameter OMP_THRESHOLD is for discontiguous tensors' operation. As for contiguous tensors' operation, the threshold is usually 10 times larger than discontiguous ones.

**Comment by zy97140** [comment]
@ezyang Could you spare some time to review this PR? I want to optimize nn module based on these macros.

**Pull request merged**